src/Components/Components/src/LayoutComponentBase.cs

```diff
@@ -17,5 +19,20 @@ public abstract class LayoutComponentBase : ComponentBase
         /// </summary>
         [Parameter]
         public RenderFragment? Body { get; set; }
+
+        /// <inheritdoc />
+        public override Task SetParametersAsync(ParameterView parameters)
+        {
+            // Derived instances of LayoutComponentBase do not appear in any statically analyzable
+            // calls of OpenComponent<T> where T is well-known. As a result, relying on reflection
+            // (ComponentProperties.SetProperties) will result in the setter being trimmed away since
+            // it appears unused. Statically reference it to ensure the setter is preserved.
+            if (parameters.TryGetValue(BodyPropertyName, out RenderFragment? body))
+            {
+                Body = body;
```

Comment:
This is a new side effect where Body is updated automatically when this is called, which I'm guessing is good/intentional?

Reply:
`base.SetParametersAsync` effectively does this, but using reflection. In fact, we're actually assigning to this property twice but doing it this way means we're not introducing any new behavior to `LayoutComponentBase`, and the assignment is statically visible to the trimmer which means it'll keep the setter if this method appears in code.

Reply:
Ah ok that makes sense then, do we still need to call base then?  Does the base do anything more than set the body?

Reply:
`SetParametersAsync` handles the entire lifetime of the component (including calling `OnInitialized` etc), so you do need to call it. Ideally we would want to avoid the reflection based assignment by doing something like this - https://docs.microsoft.com/en-us/aspnet/core/blazor/webassembly-performance-best-practices?view=aspnetcore-5.0#implement-setparametersasync-manually, but because this component can be derived, we don't know if the derived type has additional parameters that need to be assigned.

Reply:
Cool thanks for the explanation!

Reply:
@pranavkm To make the flow conceptually simpler, and avoid the double-write oddness, the pattern we've done in other cases when overriding `SetParametersAsync` is this:

```cs
public override Task SetParametersAsync(ParameterView parameters)
{
    // ... write all my own parameter properties ...

    return base.SetParametersAsync(ParameterView.Empty);
}
```

I think this is the code pattern we'd generate if we did the source-generators-for-parameter-assignment work. Might be worth going with it here too.

Reply:
Oh wait the point you make about subclassing complicates this. We'd probably be OK in the case of a layout, since there's no normal way to pass parameters to layouts. But in principle people might do weird stuff and it would be better not to block them. I guess you should keep your implementation as-is.

As for the more general case with other components and source generators, I'm not sure how we'll handle this. We *could* argue that subclasses should also override `SetParametersAsync` to assign their own derived properties, but then they can't simply pass `ParameterView.Empty` to the base.

Reply:
We might need to make the Razor compiler know whether you're directly subclassing `ComponentBase` or something else. If you are, it's safe to pass `ParameterView.Empty` to the base, and you get the perf benefit. But if your component inherits something else, we don't know whether or not to do that, and you don't get the perf benefit.

Clearly that's not a great design. We might need something more complicated.

Reply:
For the source generator, the way I had thought of it was we would code-generate some structure that would replace or feed in to [WritersForType](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/Reflection/ComponentProperties.cs#L254). We would simply replace the reflection discovery / invoke with compiled structures so we avoid some of the costs there.